File: dart/client/tests/adapter_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'conformance/adapter.dart'
    show SerializedAdapterOutput, parseAdapterListenAddress;

Future<void> main() async {
  final fixture = await _ConvexFixture.start();
  try {
    await _probeTcpAdapter(
      '0.0.0.0:19080',
      InternetAddress.loopbackIPv4,
      19080,
      fixture.url,
    );
  } finally {
    await fixture.close();
  }
  final ipv6 = parseAdapterListenAddress('[::1]:19081');
  if (ipv6.host != '::1' || ipv6.port != 19081) {
    throw StateError('adapter did not preserve bracketed IPv6 bind: $ipv6');
  }
  await _staleRelayCannotCrossAcknowledgement(replacement: false);
  await _staleRelayCannotCrossAcknowledgement(replacement: true);
  stdout.writeln('PASS Dart adapter wildcard IPv4/IPv6 TCP matrix');
}

Future<void> _staleRelayCannotCrossAcknowledgement({
  required bool replacement,
}) async {
  final server = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
  final accepted = Completer<Socket>();
  final listener = server.listen(accepted.complete);
  final clientFuture = Socket.connect(server.address, server.port);
  final peer = await accepted.future;
  final client = await clientFuture;
  final pausedAfterDequeue = Completer<void>();
  final resumeWrite = Completer<void>();
  final oldOwner = Object();
  Object? currentOwner = oldOwner;
  final writer = SerializedAdapterOutput(
    client,
    beforeWriteForTest: (event) async {
      if (event['type'] != 'subscription') return;
      pausedAfterDequeue.complete();
      await resumeWrite.future;
    },
  );
  writer.add({
    'type': 'subscription',
    'subscriptionId': 'same-id',
    'value': {'count': 0},
  }, guard: () => identical(currentOwner, oldOwner));
  await pausedAfterDequeue.future.timeout(const Duration(seconds: 1));
  currentOwner = replacement ? Object() : null;
  writer.add({'id': replacement ? 'replace' : 'unsubscribe', 'type': 'ack'});
  resumeWrite.complete();
  await writer.flush().timeout(const Duration(seconds: 1));

  final line = await peer
      .cast<List<int>>()
      .transform(utf8.decoder)
      .transform(const LineSplitter())
      .first
      .timeout(const Duration(seconds: 1));
  final event = jsonDecode(line) as Map<String, dynamic>;
  _check(
    event['type'] == 'ack' &&
        event['id'] == (replacement ? 'replace' : 'unsubscribe'),
    'stale relay crossed the ${replacement ? 'replacement' : 'unsubscribe'} acknowledgement: $event',
  );
  client.destroy();
  peer.destroy();
  await listener.cancel();
  await server.close();
}

Future<void> _probeTcpAdapter(
  String listen,
  InternetAddress connectAddress,
  int port,
  String convexUrl,
) async {
  final process = await Process.start(
    Platform.resolvedExecutable,
    ['run', 'client/tests/conformance/adapter.dart'],
    environment: {
      ...Platform.environment,
      'ADAPTER_LISTEN': listen,
      'CONVEX_URL': convexUrl,
    },
  );
  final stderrBuffer = StringBuffer();
  final listeningLine = Completer<String>();
  process.stderr.transform(utf8.decoder).transform(const LineSplitter()).listen(
    (line) {
      stderrBuffer.writeln(line);
      if (!listeningLine.isCompleted && line.contains('adapter listening')) {
        listeningLine.complete(line);
      }
    },
  );
  final listening = await listeningLine.future.timeout(
    const Duration(seconds: 10),
  );
  if (!listening.contains('adapter listening')) {
    throw StateError('adapter did not listen on $listen: $listening');
  }

  final socket = await Socket.connect(
    connectAddress,
    port,
    timeout: const Duration(seconds: 3),
  );
  final events = StreamIterator(
    socket
        .cast<List<int>>()
        .transform(utf8.decoder)
        .transform(const LineSplitter())
        .map((line) => jsonDecode(line) as Map<String, dynamic>),
  );
  try {
    await _send(socket, {'protocolVersion': 1, 'id': 'hello', 'op': 'hello'});
    final ready = await _next(events);
    _check(
      ready['type'] == 'ready' &&
          ready['id'] == 'hello' &&
          ready['language'] == 'dart' &&
          ready['protocolVersion'] == 1,
      'invalid ready event on $listen: $ready',
    );

    await _send(socket, {
      'id': 'query-success',
      'op': 'query',
      'path': 'demo:echo',
      'args': {'text': 'world'},
    });
    final success = await _next(events);
    _check(
      success['type'] == 'result' &&
          success['id'] == 'query-success' &&
          (success['value'] as Map)['text'] == 'world',
      'invalid success event: $success',
    );

    await _send(socket, {
      'id': 'query-failure',
      'op': 'query',
      'path': 'demo:fail',
      'args': <String, Object?>{},
    });
    final failure = await _next(events);
    _check(
      failure['type'] == 'error' &&
          failure['id'] == 'query-failure' &&
          (failure['error'] as Map)['name'] == 'FunctionError' &&
          ((failure['error'] as Map)['data'] as Map)['code'] ==
              'EXPECTED_FAILURE',
      'invalid structured HTTP error: $failure',
    );

    await _send(socket, {
      'id': 'subscribe',
      'op': 'subscribe',
      'subscriptionId': 'live',
      'path': 'demo:state',
      'args': {'room': 'adapter'},
    });
    final subscribeAck = await _next(events);
    final liveValue = await _next(events);
    final liveFailure = await _next(events);
    _check(
      subscribeAck['type'] == 'ack' && subscribeAck['id'] == 'subscribe',
      'invalid subscribe acknowledgement: $subscribeAck',
    );
    _check(
      liveValue['type'] == 'subscription' &&
          liveValue['subscriptionId'] == 'live' &&
          (liveValue['value'] as Map)['count'] == 0 &&
          !liveValue.containsKey('id'),
      'invalid subscription value: $liveValue',
    );
    _check(
      liveFailure['type'] == 'subscription' &&
          liveFailure['subscriptionId'] == 'live' &&
          (liveFailure['error'] as Map)['name'] == 'FunctionError' &&
          !liveFailure.containsKey('id'),
      'invalid subscription error: $liveFailure',
    );

    await _send(socket, {
      'id': 'unsubscribe',
      'op': 'unsubscribe',
      'subscriptionId': 'live',
    });
    final unsubscribeAck = await _next(events);
    _check(
      unsubscribeAck['type'] == 'ack' && unsubscribeAck['id'] == 'unsubscribe',
      'invalid unsubscribe acknowledgement: $unsubscribeAck',
    );

    await _send(socket, {'id': 'close', 'op': 'close'});
    final closed = await _next(events);
    _check(
      closed['type'] == 'closed' && closed['id'] == 'close',
      'invalid close event: $closed',
    );
  } catch (error) {
    process.kill();
    throw StateError('$error\nchild stderr:\n$stderrBuffer');
  } finally {
    await events.cancel();
  }
  await socket.close();
  final exitCode = await process.exitCode.timeout(const Duration(seconds: 3));
  if (exitCode != 0) {
    throw StateError(
      'adapter on $listen exited $exitCode\nchild stderr:\n$stderrBuffer',
    );
  }
}

Future<void> _send(Socket socket, Map<String, Object?> command) async {
  socket.writeln(jsonEncode(command));
  await socket.flush();
}

Future<Map<String, dynamic>> _next(
  StreamIterator<Map<String, dynamic>> events,
) async {
  if (!await events.moveNext().timeout(const Duration(seconds: 3))) {
    throw StateError('adapter event stream ended early');
  }
  return events.current;
}

void _check(bool condition, String message) {
  if (!condition) throw StateError(message);
}

class _ConvexFixture {
  _ConvexFixture._(this.server);

  static Future<_ConvexFixture> start() async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    final fixture = _ConvexFixture._(server);
    unawaited(server.forEach(fixture._handle));
    return fixture;
  }

  final HttpServer server;
  final List<WebSocket> _sockets = [];

  String get url => 'http://${server.address.address}:${server.port}';

  Future<void> close() async {
    for (final socket in _sockets) {
      await socket.close();
    }
    await server.close(force: true);
  }

  Future<void> _handle(HttpRequest request) async {
    if (request.uri.path == '/api/sync') {
      await _handleSync(request);
      return;
    }
    final body = jsonDecode(await utf8.decoder.bind(request).join()) as Map;
    request.response.headers.contentType = ContentType.json;
    if (body['path'] == 'demo:fail') {
      request.response.write(
        jsonEncode({
          'status': 'error',
          'errorMessage': 'expected adapter failure',
          'errorData': {'code': 'EXPECTED_FAILURE'},
          'logLines': ['adapter failure'],
        }),
      );
    } else {
      request.response.write(
        jsonEncode({
          'status': 'success',
          'value': body['args'],
          'logLines': ['adapter success'],
        }),
      );
    }
    await request.response.close();
  }

  Future<void> _handleSync(HttpRequest request) async {
    final socket = await WebSocketTransformer.upgrade(request);
    _sockets.add(socket);
    var connected = false;
    await for (final raw in socket) {
      final message = jsonDecode(raw as String) as Map<String, dynamic>;
      if (!connected) {
        connected = true;
        continue;
      }
      if (message['type'] != 'ModifyQuerySet') continue;
      final modification =
          (message['modifications'] as List).first as Map<String, dynamic>;
      if (modification['type'] != 'Add') continue;
      final queryId = modification['queryId'] as int;
      socket.add(
        jsonEncode(
          _transition(0, 1, 'AAAAAAAAAAA=', 'AQAAAAAAAAA=', queryId, {
            'type': 'QueryUpdated',
            'value': {'count': 0},
          }),
        ),
      );
      socket.add(
        jsonEncode(
          _transition(1, 1, 'AQAAAAAAAAA=', 'AgAAAAAAAAA=', queryId, {
            'type': 'QueryFailed',
            'errorMessage': 'expected subscription failure',
            'errorData': {'code': 'EXPECTED_SUBSCRIPTION_FAILURE'},
          }),
        ),
      );
    }
  }
}

Map<String, Object?> _transition(
  int startSet,
  int endSet,
  String startTimestamp,
  String endTimestamp,
  int queryId,
  Map<String, Object?> modification,
) => {
  'type': 'Transition',
  'startVersion': {'querySet': startSet, 'identity': 0, 'ts': startTimestamp},
  'endVersion': {'querySet': endSet, 'identity': 0, 'ts': endTimestamp},
  'modifications': [
    {...modification, 'queryId': queryId, 'logLines': <String>[]},
  ],
};

File: typescript/client/tests/conformance/adapter.test.ts
import assert from "node:assert/strict";
import { once } from "node:events";
import http, { IncomingMessage, Server } from "node:http";
import net from "node:net";
import { PassThrough, Readable, Writable } from "node:stream";
import test from "node:test";
import { WebSocketServer } from "ws";
import { parseListenAddress, runAdapter, startTcpAdapter } from "./adapter.js";

type Event = Record<string, any>;
type StreamAdapter = {
  input: PassThrough;
  output: PassThrough;
  send: (command: object) => boolean;
  events: (count: number) => Promise<Event[]>;
};
const zeroVersion = { querySet: 0, identity: 0, ts: "AAAAAAAAAAA=" };

test("adapter serialises success, structured HTTP failure, optional IDs, auth, and close", async () => {
  const authorizations: Array<string | undefined> = [];
  const server = http.createServer(async (request, response) => {
    const command = JSON.parse(await readBody(request));
    authorizations.push(request.headers.authorization);
    response.setHeader("content-type", "application/json");
    response.end(
      command.path === "demo:fail"
        ? JSON.stringify({
            status: "error",
            errorMessage: "adapter failure",
            errorData: { code: "ADAPTER_FAIL" },
            logLines: ["adapter log"],
          })
        : JSON.stringify({
            status: "success",
            value: { count: 3, text: "café 🦘" },
          }),
    );
  });
  const adapter = streamAdapter({ CONVEX_URL: await listen(server) });
  try {
    adapter.send({ protocolVersion: 1, id: "hello", op: "hello" });
    adapter.send({ id: "auth", op: "setAuth", token: "adapter-token" });
    adapter.send({
      id: "ok",
      op: "query",
      path: "demo:state",
      args: { nested: { value: "値" } },
    });
    adapter.send({ id: "bad", op: "mutation", path: "demo:fail", args: {} });
    adapter.input.write("not-json\n");
    adapter.send({ id: "close", op: "close" });
    const events = await adapter.events(6);
    assert.equal(events[0].type, "ready");
    assert.equal(events[0].language, "typescript");
    assert.deepEqual(events[1], { id: "auth", type: "ack" });
    assert.deepEqual(events[2], {
      id: "ok",
      type: "result",
      value: { count: 3, text: "café 🦘" },
    });
    assert.deepEqual(events[3], {
      id: "bad",
      type: "error",
      error: {
        name: "FunctionError",
        message: "adapter failure",
        data: { code: "ADAPTER_FAIL" },
      },
      logs: ["adapter log"],
    });
    assert.equal(events[4].type, "error");
    assert.equal(Object.hasOwn(events[4], "id"), false);
    assert.deepEqual(events[5], { id: "close", type: "closed" });
    assert.deepEqual(authorizations, [
      "Bearer adapter-token",
      "Bearer adapter-token",
    ]);
  } finally {
    adapter.input.destroy();
    adapter.output.destroy();
    await closeServer(server);
  }
});

test("adapter serialises subscription errors without an id and acknowledges unsubscribe", async () => {
  const server = http.createServer();
  const webSockets = new WebSocketServer({ server, path: "/api/sync" });
  webSockets.on("connection", (socket) =>
    socket.on("message", (raw) => {
      const message = JSON.parse(raw.toString());
      const add = message.modifications?.find(
        (item: Event) => item.type === "Add",
      );
      if (!add) return;
      socket.send(
        JSON.stringify({
          type: "Transition",
          startVersion: zeroVersion,
          endVersion: { querySet: 1, identity: 0, ts: "MQ==" },
          modifications: [
            {
              type: "QueryFailed",
              queryId: add.queryId,
              errorMessage: "live adapter failure",
              errorData: { code: "LIVE_ADAPTER_FAIL" },
              logLines: ["live adapter log"],
            },
          ],
        }),
      );
    }),
  );
  const adapter = streamAdapter({ CONVEX_URL: await listen(server) });
  try {
    adapter.send({
      id: "subscribe",
      op: "subscribe",
      subscriptionId: "room",
      path: "demo:fail",
      args: {},
    });
    const first = await adapter.events(2);
    assert.deepEqual(first[0], { id: "subscribe", type: "ack" });
    assert.equal(first[1].type, "subscription");
    assert.equal(first[1].subscriptionId, "room");
    assert.equal(Object.hasOwn(first[1], "id"), false);
    assert.equal(first[1].error.name, "FunctionError");
    assert.equal(first[1].error.data.code, "LIVE_ADAPTER_FAIL");
    adapter.send({
      id: "unsubscribe",
      op: "unsubscribe",
      subscriptionId: "room",
    });
    adapter.send({ id: "close", op: "close" });
    const all = await adapter.events(4);
    assert.deepEqual(all[2], { id: "unsubscribe", type: "ack" });
    assert.deepEqual(all[3], { id: "close", type: "closed" });
  } finally {
    adapter.input.destroy();
    adapter.output.destroy();
    webSockets.close();
    await closeServer(server);
  }
});

test("adapter debugDisconnect acknowledges and reconnects the active subscription", async () => {
  const server = http.createServer();
  const webSockets = new WebSocketServer({ server, path: "/api/sync" });
  let connectionCount = 0;
  webSockets.on("connection", (socket) => {
    connectionCount++;
    socket.on("message", (raw) => {
      const message = JSON.parse(raw.toString());
      const add = message.modifications?.find(
        (item: Event) => item.type === "Add",
      );
      if (!add) return;
      socket.send(
        JSON.stringify({
          type: "Transition",
          startVersion: zeroVersion,
          endVersion: {
            querySet: 1,
            identity: 0,
            ts: Buffer.from(String(connectionCount)).toString("base64"),
          },
          modifications: [
            {
              type: "QueryUpdated",
              queryId: add.queryId,
              value: { connectionCount },
            },
          ],
        }),
      );
    });
  });
  const adapter = streamAdapter({ CONVEX_URL: await listen(server) });
  try {
    adapter.send({
      id: "subscribe",
      op: "subscribe",
      subscriptionId: "room",
      path: "demo:state",
      args: {},
    });
    let events = await adapter.events(2);
    assert.equal(events[1].value.connectionCount, 1);
    adapter.send({ id: "disconnect", op: "debugDisconnect" });
    events = await adapter.events(4);
    assert.deepEqual(events[2], { id: "disconnect", type: "ack" });
    assert.equal(events[3].value.connectionCount, 2);
    adapter.send({
      id: "unsubscribe",
      op: "unsubscribe",
      subscriptionId: "room",
    });
    adapter.send({ id: "close", op: "close" });
    events = await adapter.events(6);
    assert.deepEqual(events[4], { id: "unsubscribe", type: "ack" });
    assert.deepEqual(events[5], { id: "close", type: "closed" });
  } finally {
    adapter.input.destroy();
    adapter.output.destroy();
    webSockets.close();
    await closeServer(server);
  }
});

test("ADAPTER_LISTEN serves a real TCP lifecycle without filesystem writes", async () => {
  assert.deepEqual(parseListenAddress("127.0.0.1:3210"), {
    host: "127.0.0.1",
    port: 3210,
  });
  assert.throws(() => parseListenAddress("3210"), /host:port/);
  const server = startTcpAdapter("127.0.0.1:0", {});
  await once(server, "listening");
  const address = server.address();
  if (!address || typeof address === "string")
    throw new Error("TCP adapter did not expose an address");
  const socket = net.createConnection({
    host: "127.0.0.1",
    port: address.port,
  });
  await once(socket, "connect");
  const lines = collectLines(socket);
  socket.write('{"protocolVersion":1,"id":"hello","op":"hello"}\n');
  socket.write('{"id":"close","op":"close"}\n');
  const events = await lines.waitFor(2);
  assert.equal(events[0].type, "ready");
  assert.deepEqual(events[1], { id: "close", type: "closed" });
  socket.destroy();
});

function streamAdapter(environment: NodeJS.ProcessEnv): StreamAdapter {
  const input = new PassThrough();
  const output = new PassThrough();
  const lines = collectLines(output);
  runAdapter(input, output, environment);
  return {
    input,
    output,
    send: (command) => input.write(`${JSON.stringify(command)}\n`),
    events: lines.waitFor,
  };
}
function collectLines(stream: Readable): {
  waitFor: (count: number) => Promise<Event[]>;
} {
  const values: Event[] = [];
  const waiters: Array<() => void> = [];
  let pending = "";
  stream.on("data", (chunk) => {
    pending += chunk.toString();
    const lines = pending.split("\n");
    pending = lines.pop() ?? "";
    for (const line of lines) if (line) values.push(JSON.parse(line));
    for (const waiter of waiters.splice(0)) waiter();
  });
  return {
    waitFor: async (count) => {
      const deadline = Date.now() + 2_000;
      while (values.length < count) {
        if (Date.now() > deadline)
          throw new Error("timed out waiting for adapter event");
        await new Promise<void>((resolve) => {
          waiters.push(resolve);
          setTimeout(resolve, 25);
        });
      }
      return values;
    },
  };
}
async function listen(server: Server): Promise<string> {
  server.listen({ host: "127.0.0.1", port: 0 });
  await once(server, "listening");
  const address = server.address();
  if (!address || typeof address === "string")
    throw new Error("server did not expose a TCP address");
  return `http://127.0.0.1:${address.port}`;
}
async function closeServer(server: Server): Promise<void> {
  if (!server.listening) return;
  server.close();
  await once(server, "close");
}
async function readBody(request: IncomingMessage): Promise<string> {
  let body = "";
  for await (const chunk of request) body += chunk;
  return body;
}
